### PR TITLE
Migrate to Jetpack Compose

### PR DESCRIPTION
**Pull request by Iamlooker**
Migrate:
- [x] ConsoleFragment
- [x] Migrate to Flows
- [x] ControlFragment
- [x] SettingsFragment

### CODE REVIEW

**Comment by Iamlooker** [comment]
Hi, I wanted to ask, is [this](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/build.gradle.kts#L23) intentional? Because this uses the debug signing key to sign release files

**Comment by mathiascode** [comment]
> Hi, I wanted to ask, is [this](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/build.gradle.kts#L23) intentional? Because this uses the debug signing key to sign release files

Looks like I committed it by accident after debugging an issue in the release build. It should be removed.

**Comment by Iamlooker** [comment]
> Looks like I committed it by accident after debugging an issue in the release build. It should be removed.

Done

**Comment by Iamlooker** [comment]
Btw, why are we passing StringBuilder [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)

Either way we are converting it to string as soon as we get it [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)

**Comment by Iamlooker** [comment]
```
Manifest merger failed : uses-sdk:minSdkVersion 16 cannot be smaller than version 21 declared in library [androidx.compose.material3:material3-android:1.2.1] C:\Users\LooKeR\.gradle\caches\transforms-4\5a41c055199354d74c6acce2479e2b57\transformed\material3-release\AndroidManifest.xml as the library might be using APIs not available in 16
	Suggestion: use a compatible library with a minSdk of at most 16,
		or increase this project's minSdk version to at least 21,
		or use tools:overrideLibrary="androidx.compose.material3" to force usage (may lead to runtime failures)
```

What should we do?

**Comment by Iamlooker** [comment]
Also check the warning on this [page](https://developer.android.com/jetpack/androidx/versions/all-channel) They are announcing deprecation for API 21 and below

**Comment by mathiascode** [comment]
> Btw, why are we passing StringBuilder [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)
> 
> Either way we are converting it to string as soon as we get it [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)

No good reason, we could pass a string instead.

**Comment by mathiascode** [comment]
> ```
> Manifest merger failed : uses-sdk:minSdkVersion 16 cannot be smaller than version 21 declared in library [androidx.compose.material3:material3-android:1.2.1] C:\Users\LooKeR\.gradle\caches\transforms-4\5a41c055199354d74c6acce2479e2b57\transformed\material3-release\AndroidManifest.xml as the library might be using APIs not available in 16
> 	Suggestion: use a compatible library with a minSdk of at most 16,
> 		or increase this project's minSdk version to at least 21,
> 		or use tools:overrideLibrary="androidx.compose.material3" to force usage (may lead to runtime failures)
> ```
> 
> What should we do?

We can bump the minimum version to 21. I believe the oldest Android versions have issues using the built-in Cuberite downloader anyway due to SSL errors.

**Comment by mathiascode** [comment]
I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.

**Comment by Iamlooker** [comment]
> I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.

Yes, but I did read about some changes in Android 13. I will look into this.

**Comment by Iamlooker** [comment]
Current State

**Comment by Iamlooker** [comment]
I suggest migrating LiveData to StateFlow.

This is also the suggest way to create architecture just like Jetpack Compose

**Comment by mathiascode** [comment]
> I suggest migrating LiveData to StateFlow.
> 
> This is also the suggest way to create architecture just like Jetpack Compose

If StateFlows are recommended over LiveData when using Compose, we could migrate to them. We mainly need communication between our UI and (threaded) services, which was previously achieved with LocalBroadcastManager (now deprecated).

Could the migration be done in a separate PR?

**Comment by Iamlooker** [comment]
> Could the migration be done in a separate PR?

I will open a new PR after this one

**Comment by Iamlooker** [comment]
> Could the migration be done in a separate PR?

I have been trying migration, but the livedata API seems to be unpredictable, or maybe its just because I am a little inexperienced with it. Should i do the migration is the same PR? or should I do that PR first and do Compose Migration later?

**Comment by mathiascode** [comment]
> or should I do that PR first and do Compose Migration later?

Is it possible to do before the Compose migration? If the Compose migration requires more changes anyway, it would probably be better to just do it in this PR.

**Comment by Iamlooker** [comment]
Yes, I will migrate to Flows first now :)

**Comment by Iamlooker** [comment]
Btw what is [this](https://github.com/Iamlooker/cuberite-android/blob/master/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L157) tick for?

**Comment by Iamlooker** [comment]
Can you test this? I have migrated from livedata to flows. Once all is set I will move on to compose migration

**Comment by mathiascode** [comment]
Seems to be working fine, and the changes look reasonable. Nice work!

**Comment by mathiascode** [comment]
> Btw what is [this](https://github.com/Iamlooker/cuberite-android/blob/master/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L157) tick for?

It's a hack for displaying an error when running an invalid executable. Could be replaced with something more appropriate at some point.

**Comment by Iamlooker** [comment]
This is the current state of control screen:


https://github.com/cuberite/android/assets/60404093/e2a07311-5c83-4f20-bb9e-be57f7914660

**Comment by Iamlooker** [comment]
Would you also like to migrate to [Datastore](https://developer.android.com/topic/libraries/architecture/datastore) from SharedPreferences? This is also for [reference](https://android-developers.googleblog.com/2020/09/prefer-storing-data-with-jetpack.html).

I know this is too much going on in one PR but what are your thoughts?

**Comment by mathiascode** [comment]
> This is the current state of control screen:
> 
>  sc.mp4

Could we keep UI changes in a separate PR after migration? I like the idea of displaying the IP address in the app itself too, but it would be easier to review and comment on details later.

**Comment by Iamlooker** [comment]
Cool

**Comment by Iamlooker** [comment]
I will remove these commits tomorrow

**Comment by mathiascode** [comment]
> Would you also like to migrate to [Datastore](https://developer.android.com/topic/libraries/architecture/datastore) from SharedPreferences? This is also for [reference](https://android-developers.googleblog.com/2020/09/prefer-storing-data-with-jetpack.html).
> 
> I know this is too much going on in one PR but what are your thoughts?

No strong opinion. If it's the preferred replacement, and migrating existing preferences doesn't require too much work, then we could migrate later.

**Comment by Iamlooker** [comment]
I will remove the ui commits tomorrow

**Comment by Iamlooker** [comment]
I think the removing the commits will cause some issues with my local repository. Can we merge this in this state?

**Comment by mathiascode** [comment]
Are you planning on migrating SettingsFragment in a separate PR? I'll try to review this PR next week, but I can't promise anything.

**Comment by Iamlooker** [comment]
> Are you planning on migrating SettingsFragment in a separate PR? I'll try to review this PR next week, but I can't promise anything.

If you want I can do it in the same PR

**Comment by mathiascode** [comment]
Sure, you could finish the migration in this PR if you have time.

**Comment by ItsShadowCone** [comment]
> > I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.
> 
> Yes, but I did read about some changes in Android 13. I will look into this.

Apologies if I'm completely off the track (I haven't read up on this topic in years), but last I remembered there should still be an executable-mount tmpfs that should be accessible to apps iirc. We could use this to deploy the binary from data directory there immediately before launching it and immediately after cleaning it up (Linux keeps open files around even if you remove them from the file system)

**Comment by Iamlooker** [comment]
It should be ready to review now. If I missed anything please tell me. I left the old fragments in place, because navigation to Jetpack Compose Navigation is remaining. I will start doing that once the current changes are accepted

**Comment by mathiascode** [comment]
Sorry for the delay. It will take some time to review, as I have other work taking up my time right now. I'm hoping to get this merged as soon as possible.

**Comment by mathiascode** [comment]
> > > I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.
> > 
> > 
> > Yes, but I did read about some changes in Android 13. I will look into this.
> 
> Apologies if I'm completely off the track (I haven't read up on this topic in years), but last I remembered there should still be an executable-mount tmpfs that should be accessible to apps iirc. We could use this to deploy the binary from data directory there immediately before launching it and immediately after cleaning it up (Linux keeps open files around even if you remove them from the file system)

Do you have a link to documentation regarding this? It would be nice if there was a workaround, but I haven't seen this one mentioned before.

File: app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+private fun BasicSettings(
+    onEditTheme: (DialogType) -> Unit,
+    currentTheme: Themes,
+    startOnStartup: Boolean,
+    onStartupCheck: (Boolean) -> Unit,
+    extras: @Composable CategoryScope.() -> Unit,
+) {
+    val settings = rememberCategory(
+        title = stringResource(R.string.title_activity_settings),
+        icon = Icons.Rounded.Settings,
```
Comment (mathiascode) [inline app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt:130]: Webadmin icon is currently missing. Was this icon added to the wrong category?

File: app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    }
+}
+
+@Composable
+private fun About(
+    onClick: (DialogType) -> Unit,
+) {
+    val about = rememberCategory(
+        title = stringResource(R.string.settings_info_heading),
+        footerText = stringResource(R.string.settings_info_version, BuildConfig.VERSION_NAME),
```
Comment (mathiascode) [inline app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt]: The version label should open the F-Droid page when clicked.

**Review by mathiascode — COMMENTED** [review]
Looks good in general. Seems to be working fine. I'll do some more testing on older Android versions and see if I spot any regressions.

Could you remove all unused resources, as well as the unused code in SettingsFragment? Would make it easier to see what the actual code diff is.

File: app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt (hunk 2 above)
Comment (Iamlooker) [reply app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt]: Will add this in two days, I don't have pc right now 😔

**Comment by Iamlooker** [comment]
All aforementioned issues are fixed

**Comment by Iamlooker** [comment]
Any update?

**Review by mathiascode — APPROVED** [review]

**Comment by mathiascode** [comment]
Thanks for all the great work! Are you planning on removing fragments entirely?

**Comment by Iamlooker** [comment]
Yes, I will create a PR to migrate to Compose Navigation 😉

**Pull request merged**